### PR TITLE
feat: mta.yaml and dev.mtaext validations in the problem view

### PR DESCRIPTION
**Pull request by liorfr**

### CODE REVIEW

**Review by tal-sapan — CHANGES_REQUESTED** [review]
See comments and also add tests

**Comment by tal-sapan** [comment]
The tests are failing because they don't actually run on vscode and no one bothered to add `createFileSystemWatcher`. You will have to add it to the mock.

File: src/validations/mta/mtaValidations.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  const mta = new Mta(modulePath, false, devMtaExts); // temp file is not relevant in our scenario
+
+  const validationRes: Record<string, mta.Issue[]> = await mta.validate();
+
+  for (const filePath of keys(validationRes)) {
+    diagnosticsByFile[filePath] = diagnosticsByFile[filePath] ?? [];
+
+    // iterate validation issues a single file
+    for (const issue of validationRes[filePath]) {
+      const position = new Position(issue.line, issue.column);
```
[the remark this answers is not in the document]
Comment (liorfr) [reply src/validations/mta/mtaValidations.ts:136]: nice catch!
added `convertMtaIssueCoordinateToEditorCoordinate()`

File: src/validations/mta/mtaValidations.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      diagnosticsByFile[filePath].push({
+        source: "MTA", // Should be synchronized with package.json
+        message: issue.message,
+        range: new Range(position, position),
+        severity:
+          issue.severity === "error"
```
[the remark this answers is not in the document]
Comment (liorfr) [reply src/validations/mta/mtaValidations.ts:142]: done

File: src/validations/validatoins.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { Disposable } from "vscode";
```
[the remark this answers is not in the document]
Comment (liorfr) [reply src/validations/validatoins.ts:1]: done

File: src/validations/mta/mtaValidations.ts [hunk 4]
```diff
@@ -15,40 +15,41 @@ import {
   clearDiagnosticCollections,
   getDiagnosticsCollection,
 } from "./mtaDiagnostic";
-import { DEV_MTA_EXT, MTA_YAML } from "../../utils/utils";
+
+export const MTA_YAML = "mta.yaml";
+export const DEV_MTA_EXT = "dev.mtaext";
```
Comment (tal-sapan) [inline src/validations/mta/mtaValidations.ts:20]: These variables don't have to be exported

File: src/validations/mta/mtaValidations.ts [hunk 5]
```diff
@@ -116,35 +117,47 @@ async function getMtaDiagnostics(
...
+    );
   }
 
   return diagnosticsByFile;
 }
+
+function convertMtaIssueCoordinateToEditorCoordinate(
+  coordinate: number
+): number {
+  const number = Number(coordinate) - 1;
```
Comment (tal-sapan) [inline src/validations/mta/mtaValidations.ts:158]: Why are you casting a number to number?

Comment (tal-sapan) [inline src/validations/mta/mtaValidations.ts:158]: Please add a comment that this conversion is necessary because mta issues have 1-based lines and columns (unless no line/column number is available and then it's 0), and vscode has 0-based lines and columns

File: src/validations/mta/mtaValidations.ts [hunk 6]
```diff
@@ -116,35 +117,47 @@ async function getMtaDiagnostics(
...
+      (issue) => {
+        const position = new Position(
+          convertMtaIssueCoordinateToEditorCoordinate(issue.line),
+          convertMtaIssueCoordinateToEditorCoordinate(issue.column)
+        );
+        return {
+          source: "MTA", // Should be synchronized with package.json
+          message: issue.message,
+          range: new Range(position, position),
+          severity: DiagnosticSeverity.Warning,
```
Comment (tal-sapan) [inline src/validations/mta/mtaValidations.ts:146]: mta.yaml files can have errors (unlike dev.mtaext which should only show warnings)

File: tests/mockUtil.ts [hunk 7]
```diff
@@ -1,11 +1,58 @@
...
+
+class MockRange {
+  constructor(public start: MockPosition, public end: MockPosition) {}
+}
+
+export const mockFileSystemWatcher: FileSystemWatcher = {
+  ignoreChangeEvents: false,
+  ignoreCreateEvents: false,
+  ignoreDeleteEvents: false,
+  onDidChange: (): Disposable => {
```
Comment (tal-sapan) [inline tests/mockUtil.ts:30]: Optional - if you want to make it shorter you can use
```typescript
onDidChange: (): Disposable => ({ dispose: () => undefined })
```
(For all the events in this file)

File: tests/mockUtil.ts [hunk 8]
```diff
@@ -41,6 +98,31 @@ export const testVscode: any = {
   ShellExecution: class MockShellExecution {},
   Task: class Task {},
   TaskScope: { Workspace: true },
+  languages: {
+    createDiagnosticCollection: (): DiagnosticCollection => {
+      return {
+        set: (
+          entries: ReadonlyArray<[Uri, ReadonlyArray<unknown> | undefined]>
+        ) => {
+          entries.toString();
```
Comment (tal-sapan) [inline tests/mockUtil.ts:107]: Why are you calling `toString`?

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+      // cache is empty
+      expect(keys(diagnosticCollections).length).to.equal(0);
+
+      // first call should create the diagnosticCollection
+      const firstCallDc = getDiagnosticsCollection(
+        diagnosticsName,
+        disposables
+      );
+
+      // make sure the diagnosticCollection was created and not cached
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:51]:
```suggestion
      // make sure the diagnosticCollection was created and cached
```

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+      expect(keys(diagnosticCollections).length).to.equal(1);
+
+      // second call will should get the diagnosticCollection from the cache
+      const secondCallDc = getDiagnosticsCollection(
+        diagnosticsName,
+        disposables
+      );
+
+      // reference comparison. make sure it is the same object
+      expect(diagnosticCollection === secondCallDc).to.be.true;
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:63]: Please assert object equality, it will give a better message if it fails
```suggestion
      expect(diagnosticCollection).to.equal(secondCallDc);
```
(It doesn't check deep equal unless you add `.deep`)

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+      expect(firstCallDc === secondCallDc).to.be.true;
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:64]: Same
```suggestion
      expect(firstCallDc).to.equal(secondCallDc);
```

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    beforeEach(() => {
+      clearCurrentDiagnosticCollections();
+    });
+
+    it("getDiagnosticsCollection returns cached DiagnosticCollection", async () => {
+      const diagnosticsName = "diagnosticsName";
+      const diagnosticCollection = {};
+      const createDiagnosticCollectionSpy = sinon
+        .stub(testVscode.languages, "createDiagnosticCollection")
+        .returns(diagnosticCollection);
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:40]: Instead of `returns` maybe this should do `callsFake` so a new object is returned every time. You don't really have to check later on that the returned value is what you return here, you can just check the 2 returned diagnostic collections are equal to each other.

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+      // make sure the diagnosticCollection was created and not cached
+      expect(createDiagnosticCollectionSpy.callCount).to.equal(1);
+      // cache is contain 1 diagnosticCollection
+      expect(keys(diagnosticCollections).length).to.equal(1);
+
+      // second call will should get the diagnosticCollection from the cache
+      const secondCallDc = getDiagnosticsCollection(
+        diagnosticsName,
+        disposables
+      );
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:60]: Please check `createDiagnosticCollectionSpy` is not invoked during this call. The object equality is not good enough for this check because the stub always returns the same object anyway.

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+      // reference comparison. make sure it is the same object
+      expect(diagnosticCollection === secondCallDc).to.be.true;
+      expect(firstCallDc === secondCallDc).to.be.true;
+      // cache is contain 1 diagnosticCollection
+      expect(keys(diagnosticCollections).length).to.equal(1);
+    });
+
+    it("clearDiagnosticCollections clears the collection", async () => {
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:69]:
```suggestion
    it("clearDiagnosticCollections clears the cached collections", async () => {
```

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+    it("returns the correct range when column is zero and line is 2", async () => {
+      const mtaIssue: mta.Issue = {
+        severity: "warning",
+        message: "",
+        line: 2,
+        column: 0,
+      };
+
+      const pos = {
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:119]: Please rename to `expectedPos`

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+        line: 0,
+        character: 0,
+      };
+      testGetDiagnosticsCollection(pos, mtaIssue);
+    });
+
+    it("returns the correct range when column is zero and line is 2", async () => {
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:111]: It doesn't really matter if the line is 2 or 3... what matters here is that it isn't 0
```suggestion
    it("returns the correct range when column is zero and line is non-zero", async () => {
```

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+      };
+
+      const pos = {
+        line: 1,
+        character: 0,
+      };
+      testGetDiagnosticsCollection(pos, mtaIssue);
+    });
+
+    it("returns the correct range when column and line are 3", async () => {
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:126]: Same - what matters here is that the column is non-zero
```suggestion
    it("returns the correct range when column and line non-zero", async () => {
```

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+      const mtaIssue: mta.Issue = {
+        severity: "warning",
+        message: "",
+        line: 3,
+        column: 3,
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:131]: Please use different values for the line and column so the test will check that you return them correctly (and not return `line - 1` for the column too for example)

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+      };
+
+      const pos = {
+        line: 2,
+        character: 2,
+      };
+      testGetDiagnosticsCollection(pos, mtaIssue);
+    });
+
+    function testGetDiagnosticsCollection(
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:141]: This test is for `mtaIssueToEditorCoordinate`, not `getDiagnosticsCollection`. Please rename the function and the describe.

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+        end: pos,
+      };
+
+      const resRange = mtaIssueToEditorCoordinate(mtaIssue);
+
+      expect(resRange).to.deep.equal(expectedRange);
+    }
+  });
+
+  describe("convertMtaIssueCoordinateToEditorCoordinate", async () => {
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:156]: I don't think these tests are necessary since you are already testing `mtaIssueToEditorCoordinate` with the exact same test cases

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+    ) {
+      const expectedRange = {
+        start: pos,
+        end: pos,
+      };
+
+      const resRange = mtaIssueToEditorCoordinate(mtaIssue);
+
+      expect(resRange).to.deep.equal(expectedRange);
+    }
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:153]: Please add a test for negative numbers (check it returns 0)

File: tests/validations/mta/mtaDiagnostic.spec.ts [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+
+    it("returns the correct range when column and line are 3", async () => {
+      const mtaIssue: mta.Issue = {
+        severity: "warning",
+        message: "",
+        line: 3,
+        column: 3,
+      };
+
+      const pos = {
```
Comment (tal-sapan) [inline tests/validations/mta/mtaDiagnostic.spec.ts:134]: Please rename to `expectedPos`

**Review by tal-sapan — COMMENTED** [review]
I didn't review `mtaValidations.spec.ts` yet

File: tests/mockUtil.ts (hunk 7 above)
Comment (liorfr) [reply tests/mockUtil.ts:30]: done

File: tests/validations/mta/mtaValidations.spec.ts [hunk 23]
```diff
@@ -0,0 +1,284 @@
...
+  beforeEach(() => {
+    mtaDiagnostic.clearCurrentDiagnosticCollections();
+  });
+
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  describe("watchMtaYamlAndDevExtFiles", async () => {
+    it("registers to workspace.createFileSystemWatcher and to events with the correct parameters", async () => {
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:35]: The test name should reflect what is expected instead of how it is expected to be done
```suggestion
    it("registers to changes in mta.yaml and dev.mtaext files in the workspace", async () => {
```

File: tests/validations/mta/mtaValidations.spec.ts [hunk 24]
```diff
@@ -0,0 +1,284 @@
...
+      const onDidCreateSpy = sinon.stub(mockFileSystemWatcher, "onDidCreate");
+      const onDidDeleteSpy = sinon.stub(mockFileSystemWatcher, "onDidDelete");
+
+      watchMtaYamlAndDevExtFiles(disposables);
+
+      // createFileSystemWatcher
+      expect(createFileSystemWatcherSpy.callCount).to.equal(1);
+      expect(
+        createFileSystemWatcherSpy.withArgs(`**/{mta.yaml,dev.mtaext}`).called
+      ).to.be.true;
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:49]: If you compare the arguments it will give a better error message in case of failure (otherwise it will only say "expected false to be true" which makes it hard to understand why it fails)
```suggestion
      expect(createFileSystemWatcherSpy.firstCall.args).to.deep.equal([`**/{mta.yaml,dev.mtaext}`]);
```

File: tests/mockUtil.ts (hunk 8 above)
Comment (liorfr) [reply tests/mockUtil.ts:107]: because of eslint.
I can add
`eslint-disable-next-line @typescript-eslint/no-unused-vars`

File: tests/validations/mta/mtaValidations.spec.ts [hunk 25]
```diff
@@ -0,0 +1,284 @@
...
+
+      // onDidDelete
+      expect(onDidDeleteSpy.callCount).to.equal(1);
+      expect(
+        onDidDeleteSpy.withArgs(sinon.match.func, undefined, disposables).called
+      ).to.be.true;
+    });
+
+    it("registers to workspace.onDidChangeWorkspaceFolders with the correct parameters", async () => {
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:70]: If you want, I think this description reflects the intention better:
```suggestion
    it("registers to workspace folder changes", async () => {
```

File: tests/validations/mta/mtaValidations.spec.ts [hunk 26]
```diff
@@ -0,0 +1,284 @@
...
+        Uri,
+        Diagnostic[] | undefined
+      ][] = prepDiagnostics();
+      await testUpdateMtaDiagnostics(expectedEntries);
+    });
+
+    async function testUpdateMtaDiagnostics(
+      expectedCollectionEntries: [Uri, Diagnostic[] | undefined][]
+    ) {
+      const fileName = "mtaValidations.spec.js";
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:112]: I know the function doesn't actually use the file name but please use `mta.yaml` to make it easier to understand the use case

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 12 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:40]: cache means that you get the same instance. not the same value. 
I can change it if you want bug I think this is better

File: tests/validations/mta/mtaValidations.spec.ts [hunk 27]
```diff
@@ -0,0 +1,284 @@
...
+        fsPath: uriPath,
+      } as Uri;
+
+      let collectionEntries:
+        | [Uri, Diagnostic[] | undefined][]
+        | undefined = undefined;
+      sinon.stub(testVscode.languages, "createDiagnosticCollection").returns({
+        forEach: () => {
+          return;
+        },
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:124]: I don't want to actually implement vscode APIs, but I think in this case if you want to check the `clear` logic it's necessary.
```suggestion
        forEach: (func: (uri: Uri, diagnostics: Diagnostic[] | undefined)) => {
          forEach(collectionEntries, (entry) => func(entry[0], entry[1]));
        },
```
(import `forEach` from lodash)

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 9 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:51]: changed to 
`// make sure the diagnosticCollection was created and not retrieved from cache`

File: tests/validations/mta/mtaValidations.spec.ts [hunk 28]
```diff
@@ -0,0 +1,284 @@
...
+    });
+  });
+
+  describe("updateMtaDiagnostics", async () => {
+    it("returns no diagnostics when mta.yaml does not exist", async () => {
+      await testUpdateMtaDiagnostics([]);
+    });
+
+    it("returns no diagnostics when mta.yaml and dev.mtaext have no errors", async () => {
+      sinon.stub(fsExtra, "pathExists").resolves(true);
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:95]: Please only stub this for the paths you expect

File: tests/validations/mta/mtaValidations.spec.ts [hunk 29]
```diff
@@ -0,0 +1,284 @@
...
+      ).to.be.true;
+    });
+  });
+
+  describe("updateMtaDiagnostics", async () => {
+    it("returns no diagnostics when mta.yaml does not exist", async () => {
+      await testUpdateMtaDiagnostics([]);
+    });
+
+    it("returns no diagnostics when mta.yaml and dev.mtaext have no errors", async () => {
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:94]: Add a test where mta.yaml exists and dev.mtaext doesn't exist

File: tests/validations/mta/mtaValidations.spec.ts [hunk 30]
```diff
@@ -0,0 +1,284 @@
...
+      sinon.stub(fsExtra, "pathExists").resolves(true);
+      sinon.stub(Mta.prototype, "validate").resolves({});
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:96]: Can you check the parameters sent to the `Mta` constructor?

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 13 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:60]: done

File: tests/validations/mta/mtaValidations.spec.ts [hunk 31]
```diff
@@ -0,0 +1,284 @@
...
+
+      await testUpdateMtaDiagnostics([]);
+    });
+
+    it("returns diagnostics when mta.yaml and dev.mtaext have errors", async () => {
+      const expectedEntries: [
+        Uri,
+        Diagnostic[] | undefined
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:104]: Define the return type on the signature of `prepDiagnostics` instead of in every call to the function

File: tests/validations/mta/mtaValidations.spec.ts [hunk 32]
```diff
@@ -0,0 +1,284 @@
...
+      },
+    } as Range;
+    const firstCallUri = {
+      fsPath: mtaPath,
+    } as Uri;
+    const secondCallUri = {
+      fsPath: devMtaExtPath,
+    } as Uri;
+
+    sinon.stub(fsExtra, "pathExists").resolves(true);
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:248]: Please only stub this for the expected paths

File: tests/validations/mta/mtaValidations.spec.ts [hunk 33]
```diff
@@ -0,0 +1,284 @@
...
+    sinon.stub(Mta.prototype, "validate").resolves(validationResult);
+    sinon
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:250]: Why do you need to stub this? Won't it work without the stub?

File: tests/validations/mta/mtaValidations.spec.ts [hunk 34]
```diff
@@ -0,0 +1,284 @@
...
+      .stub(mtaDiagnostic, "mtaIssueToEditorCoordinate")
+      .onFirstCall()
+      .returns(firstCallRange)
+      .onSecondCall()
+      .returns(secondCallRange);
+    sinon
+      .stub(testVscode.Uri, "file")
+      .onFirstCall()
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:258]: Instead of `onFirstCall` and `onSecondCall` stub if for the specific parameter

File: tests/validations/mta/mtaValidations.spec.ts [hunk 35]
```diff
@@ -0,0 +1,284 @@
...
+      await testUpdateMtaDiagnostics([]);
+    });
+
+    it("returns diagnostics when mta.yaml and dev.mtaext have errors", async () => {
+      const expectedEntries: [
+        Uri,
+        Diagnostic[] | undefined
+      ][] = prepDiagnostics();
+      await testUpdateMtaDiagnostics(expectedEntries);
+    });
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:107]: Add a test that checks the previous errors are cleared from the diagnostic collection (i.e. that the file paths with existing diagnostics are added with `undefined` before the new diagnostics are added)

File: tests/validations/mta/mtaValidations.spec.ts [hunk 36]
```diff
@@ -0,0 +1,284 @@
...
+        Uri,
+        Diagnostic[] | undefined
+      ][] = prepDiagnostics();
+      await testValidateWsMtaYamls(expectedEntries);
+    });
+
+    async function testValidateWsMtaYamls(
+      expectedCollectionEntries: [Uri, Diagnostic[] | undefined][]
+    ) {
+      const fileName = "mtaValidations.spec.js";
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:172]: Please use `mta.yaml` for the file name

File: tests/validations/mta/mtaValidations.spec.ts [hunk 37]
```diff
@@ -0,0 +1,284 @@
...
+          return;
+        },
+        set: (newCollectionEntries: [Uri, Diagnostic[] | undefined][]) => {
+          collectionEntries = newCollectionEntries;
+        },
+      });
+
+      await validateWsMtaYamls(disposables);
+      expect(collectionEntries).to.deep.equal(expectedCollectionEntries);
+    }
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts:195]: You should clear the diagnostics collection cache `afterEach`

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 10 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:63]: done

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 11 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:64]: done

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 14 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:69]: done

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 16 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:111]: done

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 15 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:119]: done

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 17 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:126]: done

File: tests/validations/mta/mtaValidations.spec.ts [hunk 38]
```diff
@@ -0,0 +1,284 @@
...
+  });
+
+  function prepDiagnostics() {
+    const mtaPath = resolve(__dirname, MTA_YAML);
+    const devMtaExtPath = resolve(__dirname, DEV_MTA_EXT);
+
+    const validationResult: Record<string, mta.Issue[]> = {
+      [mtaPath]: [
+        {
+          severity: "warning",
```
Comment (tal-sapan) [inline tests/validations/mta/mtaValidations.spec.ts]: There should be a test for error severity for both the mta.yaml and dev.mtaext

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 22 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:134]: done

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 18 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:131]: done

**Review by tal-sapan — COMMENTED** [review]
I didn't review the severity changes yet

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 19 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:141]: done

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 21 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:153]: done

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 20 above)
Comment (liorfr) [reply tests/validations/mta/mtaDiagnostic.spec.ts:156]: done

File: src/validations/mta/mtaValidations.ts [hunk 39]
```diff
@@ -24,44 +19,34 @@ export function watchMtaYamlAndDevExtFiles(disposables: Disposable[]): void {
...
   );
 
-  mtaFileWatcher.onDidChange(
-    (uri) => updateMtaDiagnostics(uri, disposables),
-    undefined,
-    disposables
-  );
+  const listener: (e: Uri) => Promise<void> = (uri: Uri) =>
+    updateMtaDiagnostics(uri, disposables);
+  const diagnosticParams = [undefined, disposables];
```
Comment (tal-sapan) [inline src/validations/mta/mtaValidations.ts:24]: I don't like this... these are 2 separate parameters, why are you putting them in an array and then deconstructing it everywhere?

File: tests/mockUtil.ts (hunk 8 above)
Comment (tal-sapan) [reply tests/mockUtil.ts:81]: Please disable the eslint issue or remove the parameter definition

File: tests/validations/mta/mtaDiagnostic.spec.ts (hunk 12 above)
Comment (tal-sapan) [reply tests/validations/mta/mtaDiagnostic.spec.ts:39]: `expect(A).to.equal(B)`, when A and B are objects, will check it's the same instance
https://www.chaijs.com/api/bdd/#method_equal
> Asserts that the target is strictly (===) equal to the given val.

Comment (tal-sapan) [reply tests/validations/mta/mtaDiagnostic.spec.ts:39]: You want to check that the cache is implemented in the productive function, it will be easier if you return a new object every time (then the equality assertions will fail if the cache is not implemented, which is not the case now)

File: tests/validations/mta/mtaValidations.spec.ts (hunk 28 above)
Comment (liorfr) [reply tests/validations/mta/mtaValidations.spec.ts:95]: done

File: tests/validations/mta/mtaValidations.spec.ts (hunk 37 above)
Comment (liorfr) [reply tests/validations/mta/mtaValidations.spec.ts:195]: done

File: tests/validations/mta/mtaValidations.spec.ts (hunk 33 above)
Comment (liorfr) [reply tests/validations/mta/mtaValidations.spec.ts:250]: done

**Review by liorfr — COMMENTED** [review]
unit tests added in
https://github.com/SAP/vscode-mta-tools/pull/167
